File: Two_Pointers/Patterns/Pattern4/Pattern4.java
import java.util.*;
public class Pattern4 {
    public static void main(String[] args) {
        String s="asdfghjklal";
    Map<Character,Integer> map=new HashMap<>();
    int left=0;
    int right=0;
    int maxx=0;

    while(right<s.length() && right>=left){
        char c=s.charAt(right);
        map.put(c,map.getOrDefault(c,0)+1);

        while(map.get(c)>1){
            char d=s.charAt(left++);
            map.put(d,map.get(d)-1);
            if(map.get(d)==0){
                map.remove(d);
            }
        }
        maxx=Math.max(maxx,right-left+1);
        right++;

    }
    System.out.println(maxx);
    }
    
}

File: Two_Pointers/Patterns/Pattern4/Pattern4.py
s="helloworld"
dct={}
left=0
right=0
maxx=0

while right<len(s):
    c=s[right]
    dct[c]=dct.setdefault(c,0)+1
    while dct[c]>1:
        d=s[left]
        dct[d]=dct.get(d)-1
        left+=1
    maxx=max(maxx,right-left+1)
    right+=1
print(maxx)